File: src/lib/protocol/tags.ts
import type { NostrEvent, NostrTag } from './event';

export type IndexedTags = {
  readonly events: readonly string[];
  readonly pubkeys: readonly string[];
  readonly topics: readonly string[];
  readonly relays: readonly string[];
};

export function firstTagValue(
  event: NostrEvent,
  name: string,
): string | undefined {
  return event.tags.find((tag) => tag[0] === name)?.[1];
}

export function tagValues(event: NostrEvent, name: string): string[] {
  return event.tags.flatMap((tag) =>
    tag[0] === name && tag[1] ? [tag[1]] : [],
  );
}

export function indexTags(event: NostrEvent): IndexedTags {
  return {
    events: tagValues(event, 'e'),
    pubkeys: tagValues(event, 'p'),
    topics: tagValues(event, 't'),
    relays: tagValues(event, 'r'),
  };
}

export function replyRoot(event: NostrEvent): string | undefined {
  return marker(event.tags, 'root') ?? tagValues(event, 'e')[0];
}

export function replyParent(event: NostrEvent): string | undefined {
  const reply = marker(event.tags, 'reply');
  if (reply) return reply;
  const events = tagValues(event, 'e');
  return events.length > 1 ? events[events.length - 1] : undefined;
}

function marker(
  tags: readonly NostrTag[],
  markerName: string,
): string | undefined {
  return tags.find((tag) => tag[0] === 'e' && tag[3] === markerName)?.[1];
}

File: crates/lkjstr-protocol/src/tags.rs
use serde::{Deserialize, Serialize};

use crate::NostrEvent;

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct IndexedTags {
    pub events: Vec<String>,
    pub pubkeys: Vec<String>,
    pub quotes: Vec<String>,
    pub addresses: Vec<String>,
    pub topics: Vec<String>,
    pub relays: Vec<String>,
}

pub fn first_tag_value(event: &NostrEvent, name: &str) -> Option<String> {
    event
        .tags
        .iter()
        .find(|tag| tag.first().is_some_and(|item| item == name))
        .and_then(|tag| tag.get(1))
        .cloned()
}

pub fn tag_values(event: &NostrEvent, name: &str) -> Vec<String> {
    event
        .tags
        .iter()
        .filter(|tag| tag.first().is_some_and(|item| item == name))
        .filter_map(|tag| tag.get(1).filter(|value| !value.is_empty()).cloned())
        .collect()
}

pub fn index_tags(event: &NostrEvent) -> IndexedTags {
    IndexedTags {
        events: tag_values(event, "e"),
        pubkeys: tag_values(event, "p"),
        quotes: tag_values(event, "q"),
        addresses: tag_values(event, "a"),
        topics: tag_values(event, "t"),
        relays: tag_values(event, "r"),
    }
}

pub fn reply_root(event: &NostrEvent) -> Option<String> {
    marker(&event.tags, "root").or_else(|| tag_values(event, "e").first().cloned())
}

pub fn reply_parent(event: &NostrEvent) -> Option<String> {
    if let Some(reply) = marker(&event.tags, "reply") {
        return Some(reply);
    }
    let events = tag_values(event, "e");
    if events.len() > 1 {
        events.last().cloned()
    } else {
        None
    }
}

fn marker(tags: &[Vec<String>], marker_name: &str) -> Option<String> {
    tags.iter()
        .find(|tag| {
            tag.first().is_some_and(|item| item == "e")
                && tag.get(3).is_some_and(|item| item == marker_name)
        })
        .and_then(|tag| tag.get(1))
        .cloned()
}
